### PR TITLE
refactor(client): split lower-jaw component to subcomponents

### PR DESCRIPTION
**Pull request by Sboonny**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Note: I have spent a long while refactoring and haven't tested it yet. Nothing should be changed, I didn't try to change style or the logic behind the function.

Why turn the element to jsx.element.

- jsx.element has its own state, which means when the camper interacts with the component, the parent doesn't re-render.
- when the parent state changes, it doesn't need to render the elements again, because it remembers the state of jsx.element.
- It's way easier to understand and read.

*I will open it for review after testing it 👍*

### CODE REVIEW

File: cypress/e2e/default/learn/challenges/multifile.ts
Comment (Sboonny) [inline cypress/e2e/default/learn/challenges/multifile.ts:7]:
```suggestion
  lowerJawButton: '[data-cy=lowerJaw-button]',
```

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (bbsmooth) [inline client/src/templates/Challenges/classic/lower-jaw.tsx:139]: `testScreenReaderText` doesn't seem to be defined in translations.json. The original sr-only text used here was `aria.submit` ("Use Ctrl + Enter to submit") and was only used when the keyboard focus was still in the editor (the assumption being that the user used Ctrl + Enter to run the tests and thus they would also use Ctrl + Enter to submit their answer). I suppose it wouldn't hurt to show it even if the focus is on the submit button, since you can use use Ctrl + Enter. Or we could remove it completely since screen reader users probably don't need to keep hearing this over and over. But if we did that then it would be nice to document the Ctrl + Enter behavior somewhere so screen reader users will know that it exits.

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:139]: > ("Use Ctrl + Enter to submit") and was only used when the keyboard focus was still in the editor (the assumption being that the user used Ctrl + Enter to run the tests and thus they would also use Ctrl + Enter to submit their answer). 

There is slight confusion on my part, was it originally part of editor? meaning when screen reader focus on the editor, they will hear "Use Ctrl + Enter to submit".

If so, I like that more than the current one, because while navigating to the lower jaw, I was pressing tab multiple times which created tabs in the editor, then I had to press "Ctrl + Enter" to move to the confirm button.

Comment (bbsmooth) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:139]: > was it originally part of editor? meaning when screen reader focus on the editor, they will hear "Use Ctrl + Enter to submit".

I don't think so. I'm not hearing any such thing in the production version right now. Currently, the only time this text is heard is when the user uses Ctrl + Enter to check their code and it passes. Then we tack on this text to remind them that they can use Ctrl + Enter to submit as well.

> I was pressing tab multiple times which created tabs in the editor

Yes, by default the Tab key is trapped in the editor and you have to use Ctrl + m to "free" it so that you can then Tab to the button below the editor. This is difficult to know about since we don't have any documentation about it.

> Edit: if this isn't the case, maybe we should add screen reader text for keyboard shortcuts' shortcut

This isn't just a screen reader issue, so the solution should not involve sr-only text. The majority of people using this site with the keyboard will not be using a screen reader. So if we need to make sure we have a way to communicate to ALL keyboard users what the shortcuts are. We are getting a little off topic here. I think I might open an issue about this.

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:139]: Ya, this seems out of the scope of the PR 👍 🤔. I think it's better for now to add the original text, and we can replace the text with better communication to the keyboard users. 👍

File: client/src/templates/Challenges/components/lower-jaw-button.tsx
Comment (Sboonny) [inline client/src/templates/Challenges/components/lower-jaw-button.tsx:36]: There is something weird happening, that stops the browsers from focusing on this button, when I use `focus-visible` to set the focus.

Normally browsers are smart enough to do focus on the button, when the function triggers.

Comment (bbsmooth) [reply client/src/templates/Challenges/components/lower-jaw-button.tsx:36]: @Sboonny Can you be a little more specific about what you are having a problem with. I just tested this and it seems to be working fine for me. If the keyboard focus is on the Check Your Code button and you either use the Enter key/Space bar to click the button or use Ctrl + Enter then the focus remains on the button when it switches to Submit. But if the keyboard focus is in the editor and you use Ctrl + Enter then the focus stays in the editor and does not switch to the button, which is how it should be. In other words, if keyboard focus is on the button then it stays on the button, and if keyboard focus is in the editor then it stays in the editor.

Also, the `:focus-visible` pseudo-class doesn't have anything to do with setting focus. It just allows you to display the keyboard focus outline only when the keyboard is actually being used.

Comment (Sboonny) [reply client/src/templates/Challenges/components/lower-jaw-button.tsx]: > Also, the `:focus-visible` pseudo-class doesn't have anything to do with setting focus. It just allows you to display the keyboard focus outline only when the keyboard is actually being used.

I was confused by the below paragraph in this article. https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html

> “Mode of operation” accounts for user agents which may not always show a focus indicator, or only show the focus indicator when the keyboard is used. User agents may optimise when the focus indicator is shown, such as only showing it when a keyboard is used. Authors are responsible for providing at least one mode of operation where the focus is visible. In most cases there is only one mode of operation so this success criterion applies. The focus indicator must not be time limited, when the keyboard focus is shown it must remain.

I was like, "Event is part of the button, right? So when someone interacts with the event using the keyboard, then they are interacting with the button and will show focus on that button". 🤦 

Reading your comment, I was just hoping it would trigger, I should have created a quick code pen to test this first, instead of assuming. 😅 

---

> @Sboonny Can you be a little more specific about what you are having a problem with. I just tested this and it seems to be working fine for me. If the keyboard focus is on the Check Your Code button and you either use the Enter key/Space bar to click the button or use Ctrl + Enter then the focus remains on the button when it switches to Submit. But if the keyboard focus is in the editor and you use Ctrl + Enter then the focus stays in the editor and does not switch to the button, which is how it should be. In other words, if keyboard focus is on the button then it stays on the button, and if keyboard focus is in the editor then it stays in the editor.

Because of the existence of the tests, I assumed that the accessible behavior, which reminds me of the talk we had about embedding `input` element inside a `button`, which is a WCAG violation and thus not accessible. 

So I really should start looking into component, and see if it's actual accessible and not assume. I should have learned that in that issue 🤦 😥

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/classic/lower-jaw.tsx:124]: This is a strange pattern - it makes it unclear what the text belongs to or what the click listeners are for.

Essentially what we have here is a component that contains some buttons.  Why not just define the three types of buttons as components and then conditionally render them here?

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:124]: I have done so, to preserve the focus state on the button.

Here is its test

```ts
  it('focuses the submit button after testing a valid solution', () => {
    cy.visit(location);
    cy.focused().type('{end}{enter}<meta charset="UTF-8" />');
    cy.get(selectors.lowerJawButton).should('not.be.focused');
    cy.get(selectors.lowerJawButton).click();
    cy.get(selectors.lowerJawButton).should('be.focused');
  });
  ```
  
 I think the intended behavior is for the camper to focus on the button, run the "test the solution" function, and be able to press space or enter again to start the submit function, without the need to navigate to the button with keyboard.

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:124]: Should I turn this into variables for readability? or create two buttons and use `useRef` to set the focus?

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:124]: Right, but in production the screenreader announces that the submit button exists when you click on "Check Your Code". That doesn't happen in this PR, presumably because the text is modified instead of the button as a whole being replaced.

Since this is supposed to be a refactor, it should not change that behaviour. So, yes, I think two buttons + useRef is the way to go.

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx]: It worked without `useRef`, I still need to learn more about how the button are parsed/rendered

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (Sboonny) [inline client/src/templates/Challenges/classic/lower-jaw.tsx:9]:
```suggestion
import { MAX_MOBILE_WIDTH } from '../../../../../config/misc';
import { apiLocation } from '../../../../../config/env.json';
```

**Review by Sboonny — COMMENTED** [review]
Clean the changes a little

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (Sboonny) [inline client/src/templates/Challenges/classic/lower-jaw.tsx:3]:
```suggestion
import { Button } from '@freecodecamp/react-bootstrap';

```

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (ojeytonwilliams) [reply client/src/templates/Challenges/classic/lower-jaw.tsx]: It keeps focus now, which is good, but it's still not the same as production.

In production, when you click the "Check Your Code" button it announces that there's a "Submit and go to the next challenge" button on the page. That's no longer the case in this PR.

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx]: I have tried to record my narrator reading the buttons, but couldn't because the extension doesn't catch system sounds, even after allowing it to.

I can't replicate the issue of it not reading "Submit and go to the next challenge"

Comment (ojeytonwilliams) [reply client/src/templates/Challenges/classic/lower-jaw.tsx]: I think you should find a way to browse the site with a screen-reader since it's important to not accidentally damage it's accessibility.  One option would be to use Orca inside WSL.  That should give a better dev experience than Windows given how little we support it.

Regardless, the issue still exists, so please revert the changes to `setTestBtnAriaHidden`.

Comment (shootermv) [reply client/src/templates/Challenges/classic/lower-jaw.tsx]: can i suggest better title? may it be "split lower-jaw component to subcomponents"?

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx]: > Regardless, the issue still exists, so please revert the changes to `setTestBtnAriaHidden`.

If the issue is persisting, we need to add keys to the component, so it renders a different state, then add the logic to focus on it when the challenge is complete, and make sure that it doesn't focus on when someone press `Ctrl+enter` while there aren't in the lower jaw

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (Sboonny) [inline client/src/templates/Challenges/classic/lower-jaw.tsx:59]:
```suggestion
      setTestBtnAriaHidden(false);
      setIsFeedbackHidden(false);
```

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (Sboonny) [inline client/src/templates/Challenges/classic/lower-jaw.tsx:48]: Is it actually confused by generic?

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx]: > If the issue is persisting, we need to add keys to the component, so it renders a different state, then add the logic to focus on it when the challenge is complete, and make sure that it doesn't focus on when someone press `Ctrl+enter` while there aren't in the lower jaw.

I just noticed that this can be misunderstood, I have implemented the change, if it's an issue I can do more.

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
@bbsmooth Just confirming, would like to add any further requests (I did not see any blocking "request for change" earlier)?

**Comment by bbsmooth** [comment]
@raisedadead @Sboonny I'm trying to test this PR but am not having any luck getting it running, so I don't know if I have any other concerns at the moment. I'll try to get help with this in Discord. But if it looks good to you and others then don't wait for me.

**Comment by bbsmooth** [comment]
@Sboonny @raisedadead I think I have a good fix in place that will ensure screen readers announce the Submit button. I have tested it with NVDA, JAWS (both Windows screen readers), and Orca (linux screen reader) and am getting very good results. There are too many changes to describe here. Would it be OK if I pushed my fixes to this PR?

**Comment by Sboonny** [comment]
Yeab 👍, feel free to push it

**Comment by bbsmooth** [comment]
I've pushed my changes. Please give it a try and verify that everything seems to be working. I don't have time at the moment but I will explain my reasoning for some of these changes tomorrow.

File: client/src/templates/Challenges/classic/lower-jaw.tsx
Comment (Sboonny) [inline client/src/templates/Challenges/classic/lower-jaw.tsx:285]: Could we use `aria-relevant` in submit button 🤔, to announce the change in the submit button component, instead of relaying on ref?

Comment (bbsmooth) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:285]: `aria-relevant` is only used for live regions so it won't have any affect on the buttons. The refs on the Check Your Code and Submit buttons are being used for two things:
- To set focus on the Submit button
- To determine whether the Check Your Code button was clicked
I suppose we could give each of these buttons an `id` instead and use `getElementById` to do these two things, but I was under the impression that using `ref` to access these buttons is the proper way in react.

Comment (Sboonny) [reply client/src/templates/Challenges/classic/lower-jaw.tsx:285]: Yeab it's the proper way to do it in react 👍, I was greedy to know how it would function with aria-relevant.

**Review by bbsmooth — COMMENTED** [review]
I am satisfied with how this works with the NVDA and JAWS screen readers on Win 11 and the Orca screen reader on Linux. I don't have a Mac so I can only assume it works adequately with VoiceOver (and I have no reason to believe it wouldn't), but if we have someone who is capable of testing with VO that would be great. The primary tests are:

- Use `Ctrl + Enter` while focus is in editor to trigger a code check: 
  - If code does not pass then the error message and hint should be announced automatically. Keyboard focus should remain in the editor.
  - If code passes then the congrats message should be announced automatically with the addition of telling the user they can use `Ctrl + Enter` to submit. Keyboard focus should remain in editor.

- Use the Check Your Code button to trigger code check: 
  - If code does not pass then error message and hint should be read automatically and focus should stay on Check Your Code button.
  - If code passes then congrats message should be read automatically followed by announcing the "Submit and go to next challenge" button and focus should be on button. If the button is announced before the congrats message this is still a pass. If the button is not announced at all this is a fail.

On an unrelated note, I am able to consistently trigger an Unhandled Runtime Error with the following steps:

- Open any Step with the browser wide enough to be in desktop layout
- Do a hard refresh (`Shift` + reload button)
- Narrow the browser as skinny as it will go very quickly. At some point an Unhandled Runtime Error dialog will pop-up.

I am able to trigger this in FF on both Linux and Win 11. On Win 11 I usually have to narrow/expand the browser several times before it triggers. On Linux it usually happens the first time. I do not seem to be able to trigger this with Chrome. This is the first time I've seen this, so at first I thought it might be an issue with this PR but now I am thinking that this is probably something else that perhaps only affects FF. I am not able to trigger this on production, so perhaps there is something unique to the development environment that is causing this?

Here's a gif just so you all don't think I'm seeing things.

simplescreenrecorder

**Comment by Sboonny** [comment]
> I am able to trigger this in FF on both Linux and Win 11. On Win 11 I usually have to narrow/expand the browser several times before it triggers. On Linux it usually happens the first time. I do not seem to be able to trigger this with Chrome. This is the first time I've seen this, so at first I thought it might be an issue with this PR but now I am thinking that this is probably something else that perhaps only affects FF. I am not able to trigger this on production, so perhaps there is something unique to the development environment that is causing this?

Yeah, the error originates from `.cashe` folder, running `pnpm run clean-and-develop` will hopefully sort it. 

Can you share full error, even if running the command solve it, I want to investigate it more

**Review by bbsmooth — APPROVED** [review]
Sorry, I didn't mean to bog down this PR with a most likely unrelated error. Let's forget about this for now and if I experience it further I will open up a proper issue.

I think it would be great if we have someone who is capable of using VoiceOver to test this, but if not then I don't think it should hold up this PR. I see no reason why it wouldn't work properly with VO, but once in a while you can get a weird "bug" with a specific screen reader.

So this has my blessing as is.

**Pull request merged**
